u2f-ref-code/java/src/com/google/u2f/key/messages/TransferAccessMessage.java

```diff
@@ -0,0 +1,129 @@
+/**
+ * 
+ */
+package com.google.u2f.key.messages;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Arrays;
+import java.util.Objects;
+
+import com.google.u2f.U2FException;
+
+/**
+ * @author at
+ */
+public class TransferAccessMessage {
+  private final int sequenceNumber;
+  private final byte[] newUserPublicKey;
+  private final byte[] applicationParameter;
+  private final X509Certificate attestationCertificate;
+  private final byte[] signatureAuthenticationKey;
+  private final byte[] signatureAttestationKey;
+
+  public TransferAccessMessage(int sequenceNumber, byte[] newUserPublicKey,
+      byte[] applicationParameter, X509Certificate attestationCertificate,
+      byte[] signatureAuthenticationKey, byte[] signatureAttestationKey) {
+    this.sequenceNumber = sequenceNumber;
+    this.newUserPublicKey = newUserPublicKey;
+    this.applicationParameter = applicationParameter;
+    this.attestationCertificate = attestationCertificate;
+    this.signatureAuthenticationKey = signatureAuthenticationKey;
+    this.signatureAttestationKey = signatureAttestationKey;
+  }
+  
+  public static TransferAccessMessage fromByte(byte[] data) throws U2FException {
+    try {
+      DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(data));
+
+      int sequenceNumber = inputStream.readUnsignedByte();
+      byte[] oldUserPublicKey = new byte[65];
+      inputStream.readFully(oldUserPublicKey);
+      byte[] newUserPublicKey = new byte[65];
+      inputStream.readFully(newUserPublicKey);
+      byte[] applicationParameter = new byte[32];
+      inputStream.readFully(applicationParameter);
+      X509Certificate attestationCertificate = (X509Certificate) CertificateFactory.getInstance(
+          "X.509").generateCertificate(inputStream);
+      byte[] signatureAuthenticationKey = new byte[inputStream.readUnsignedByte()];
+      inputStream.readFully(signatureAuthenticationKey);
+      byte[] signatureAttestationKey = new byte[inputStream.available()];
+      inputStream.readFully(signatureAttestationKey);
+      
+      if (inputStream.available() != 0) {
+        throw new U2FException("Message ends with unexpected data");
```

Comment:
well, this should not be possible, given that the line above it reads inputStream.available(). Should I change the signatureUsingAttestationKey to work the same as the signature before it? (Use length field to determine how many bytes to read instead of reading the rest? This costs us a byte for the length field.